forem/forem
# Override Tag Moderation Actions

## What type of PR is this? (check all applicable)


## Description

In this PR we: 

1. Remove the restriction that prevents tag moderators from adding or removing a tag that was added or removed by another tag moderator.
2.  Remove the restriction that prevents admins from adding or removing a tag that was added or removed by another admin
3.  The previous requirement was ~'When a tag moderator attempts to add or remove a tag that was added or removed by an admin, the tag moderation panel should show a warning that can be dismissed, stating, "Tag has been {added/removed} by Admin. You can't {add/remove} it now. If you feel that's a mistake, please contact us at {support_email}".'~ However, after chatting with Phil we will proactively show a message "Tag has been {added/removed} by Admin. You can't {add/remove} it now. If you feel that's a mistake, please contact us at {support_email}" when a tag moderator is moderating a tag that has already been updated by an admin or super moderator. 

## Related Tickets & Documents

- Related Issue #https://github.com/forem/forem/issues/19824
- Closes #19824 

## QA Instructions, Screenshots, Recordings

### QA instructions 

The Qa instructions are length and can be found in the demo, hence I would advise for more thorough instructions to watch the video (preferable on 1.5x speed)

A. Create an article and publish.

B. Create users: 
- User 1: should be an admin
- User 2: should be an admin 
- User 3: should be a tag moderator for #javascript (or another tag of your choice)
- User 4: should be a tag moderator for #javascript (or another tag of your choice)
 
C. You can update users permissions on /admin/member_manager/users
D. You can update the tags a tag moderator will moderate on /admin/content_manager/tags/<tag>

**To test Use Case (1): Remove the restriction that prevents tag moderators from adding or removing a tag that was added or removed by another tag moderator.**
- Let user 3 add #javascript
- Ensure that user 4 can remove #javascript

**To test Use Case (2): Remove the restriction that prevents admins from adding or removing a tag that was added or removed by another admin.**
- Let user 1 add #career
- Let user 2 remove #career 

**To test Use Case (3): However, after chatting with Phil we will proactively show a message "Tag has been {added/removed} by Admin. You can't {add/remove} it now. If you feel that's a mistake, please contact us at {support_email}" when a tag moderator is moderating a tag that has already been updated by an admin or super moderator.** 
- Let user 1 add/remove #javascript 
- You should see a message that user 3 cannot add/remove the tag when navigating to the mod panel for the article from their profile.. 


### Loom Video

**Tiered Moderators and Admins: Testing Overrides and Tag Adjustments:**

https://www.loom.com/share/3c0ba410ef3b49bfbaf3f56d86fea9a8

Note: we allow admins and super moderators to create a tag adjustment for a tag that was adjusted by another admin. I forgot to mention super moderators.

In addition, I've made the fix to the UI that I pointed out: 

<img width="361" alt="Screenshot 2023-09-12 at 13 26 54" src="https://github.com/forem/forem/assets/2786819/25d02785-3e9f-48cb-aab5-f97f8a5c0360">

**Authors not allowed to add/remove tags that were previously adjusted by moderators and admins**

https://www.loom.com/share/e1b0b51930934e3ebcd40f7a76a40e26


Test the same behaviour on /mod. 

### UI accessibility checklist
N/A

## Added/updated tests?

      have not been included_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/fItgT774J3nWw/giphy.gif)
